mutable_buffer/src/database.rs

```diff
@@ -199,6 +211,47 @@ impl MutableBufferDb {
 
         size
     }
+
+    /// Returns the partitions in the requested sort order
+    pub fn partitions_sorted_by(
+        &self,
+        sort_rules: &PartitionSortRules,
+    ) -> Vec<Arc<RwLock<Partition>>> {
+        let partitions: Vec<_> = {
+            let partitions = self.partitions.read().expect("poisoned mutex");
+            partitions.values().cloned().collect()
+        };
+
+        let mut partitions: Vec<_> = match &sort_rules.sort {
+            PartitionSort::CreatedAtTime => {
+                let times: Vec<_> = partitions
+                    .iter()
+                    .map(|p| p.read().expect("mutex poisoned").created_at)
+                    .collect();
+                let mut pt: Vec<_> = partitions.into_iter().zip(times).collect();
```

Comment:
It might make sense You could probably map to `(partition, partition.created_at)` in the same loop if you wanted to

```suggestion
                let mut pt: Vec<_> = partitions
                    .into_iter()
                    .map(|p| {
                      let created_at = p.read().expect("mutex poisoned").created_at
                      (p, created_at)
                    })
                    .collect();
```

I don't see any big difference between this and what you have, though